master/internal/task/allocation.go

```diff
@@ -171,224 +206,353 @@ func NewAllocation(
-			idle.Unregister(cfg.ServiceID)
+	switch msg := msg.(type) {
+	case *sproto.ResourcesAllocated:
+		if err := a.resourcesAllocated(msg); err != nil {
+			a.crash(err)
 		}
-		allocationmap.UnregisterAllocation(a.model.AllocationID)
-	case sproto.ContainerLog:
+	case *sproto.ResourcesStateChanged:
+		a.resourcesStateChanged(msg)
+	case *sproto.ResourcesFailure:
+		a.restoreResourceFailure(msg)
+	case *sproto.ReleaseResources:
+		a.releaseResources(msg)
+	case *sproto.ContainerLog:
 		a.sendTaskLog(msg.ToTaskLog())
+	case *sproto.InvalidResourcesRequestError:
+		a.crash(msg.Cause)
+	default:
+		panic(fmt.Errorf("unexpected RM event"))
+	}
+}
 
-	// These messages allow users (and sometimes an orchestrator, such as HP search)
-	// to interact with the allocation. The usually trace back to API calls.
-	case AllocationReady:
-		// AllocationReady only comes from the running container, so to
-		// avoid a race condition with the slower transition to running state
-		// which comes via polling for dispatcher RM, move the state to running now.
-		a.setMostProgressedModelState(model.AllocationStateRunning)
-		a.model.IsReady = ptrs.Ptr(true)
-		if err := a.db.UpdateAllocationState(a.model); err != nil {
-			a.Error(ctx, err)
-		}
-		a.sendTaskLog(&model.TaskLog{Log: fmt.Sprintf("Service of %s is available", a.req.Name)})
-	case AllocationWaiting:
-		a.setMostProgressedModelState(model.AllocationStateWaiting)
-		if err := a.db.UpdateAllocationState(a.model); err != nil {
-			a.Error(ctx, err)
-		}
-	case MarkResourcesDaemon:
-		if err := a.SetResourcesAsDaemon(ctx, msg.AllocationID, msg.ResourcesID); err != nil {
-			a.Error(ctx, err)
-		}
-	case sproto.AllocationSignal:
-		a.HandleSignal(ctx, sproto.AllocationSignalWithReason{AllocationSignal: msg})
-	case sproto.AllocationSignalWithReason:
-		a.HandleSignal(ctx, msg)
-	case AllocationState:
-		if ctx.ExpectingResponse() {
-			ctx.Respond(a.State())
-		}
-	case SetAllocationProxyAddress:
-		if len(a.req.ProxyPorts) == 0 {
-			if ctx.ExpectingResponse() {
-				ctx.Respond(ErrBehaviorUnsupported{Behavior: fmt.Sprintf("%T", msg)})
-			}
-			return nil
-		}
-		a.model.ProxyAddress = &msg.ProxyAddress
-		if err := a.db.UpdateAllocationProxyAddress(a.model); err != nil {
-			a.Error(ctx, err)
+// State returns a copy of the current State of the allocation.
+func (a *allocation) State() AllocationState {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.state()
+}
+
+// IsRestoring returns if the allocation has been restored by the resource manager.
+func (a *allocation) IsRestoring() bool {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.req.Restore && !a.restored
+}
+
+// waitForRestore waits until the allocation has been restored by the resource manager or a minute
+// has passed. If a minute passes, an error is returned. The allocation must exist otherwise this
+// will return a not found error.
+func (a *allocation) waitForRestore(ctx context.Context) error {
+	for i := 0; i < 60; i++ {
```

Comment:
Hmm, this waits for sixty iterations, but has no idea how long that's taken. If, for example, it blocks at `a.IsRestoring()` for awhile, it could wait much longer (which maybe is fine). Something like the following would be more accurate, but maybe that doesn't matter.
```go
waitStart := time.Now()
for ... {
    switch {
    //...
    }
    if time.Since(waitStart) > time.Minute {
        break
    } else {
        time.Sleep(time.Second)
    }
}
```

Reply:
copy pasta'd from old code, but yeah. going to just wrap the ctx with a timeout and use time.Ticker